Filesize Reporter module

petergoes opened the pull request:
Move the reporting to the user outside of the core module. The CLI file takes the data that is returned by the module and feeds that to a new reporter module. In the future, (#75) this includes error reporting as well.

A few modules needed refactoring to make this happen:
- The loglevel in the `logger` can be updated during runtime
- The `parser-stats` module is removed in favour of this new module
- The index, exposes the `src` and `dest` folder from the config setting, we need the `dest` folder to let the user know where to find the modified files

petergoes committed:
refactor(parse-all): report error per parser

When a parser errors, report what happend to the user. Continue with all
parsers before exiting. Parsers are started in paralell, we need to wait
for all of them to complete to clean up the temp folder.

petergoes committed:
refactor(index): return the promise chain

By returning the promise chain consuming modules can act on the
resolving or rejecting of the promise chain. As a result, the
bin/fastatic module can set the exitCode as discussed in #75

petergoes committed:
Merge branch 'master' into fix/exit-process-after-error

petergoes committed:
Merge branch 'master' into fix/exit-process-after-error

petergoes committed:
refactor(index): split execution from cleanup flow

To be able to return the result to consuming modules, the result of the
parsing should be returned before the cleanup is executed

petergoes committed:
feat(results): add results module

The results module returns an object with stats. The fastatic module
returns this object to consuming modules

petergoes committed:
refactor(results): split functions in smaller pieces

petergoes committed:
Merge branch 'master' into output-result-as-data

petergoes committed:
refactor(results): remove saved bytes from result

petergoes committed:
refactor(results): rename to compareFileSize

petergoes committed:
refactor(compare-file-size): accept two dirs instead of config object

petergoes committed:
refactor(compare-file-size): return sizes object per folder

petergoes committed:
refactor(compare-file-size): put calculating file sizes in own module

petergoes committed:
refactor(index): return an object with the fileSizes as a property

Now we can output more data to consuming modules

petergoes committed:
feat(reporter): add reporter skeleton module

petergoes committed:
fix(index): replace console log with the logger module

petergoes committed:
refactor: create temp folders for src and dist files

We copy the src users source folder to a temp dir. We do that so we can
compare the files in a time when the src folder might be overridden by
the parsed files

petergoes committed:
refactor(index): remove unneeded Promise.all

petergoes committed:
refactor(compare-file-sze): use path module to append dir to files

petergoes committed:
refactor(get-filesizes-per-type): simplify variable names

petergoes committed:
refactor(get-filesizes-per-type): simplify size addition

petergoes committed:
refactor(parse-all): remove unused dependencies

petergoes committed:
fix(get-filesizes-per-type): fix variable naming collisions

petergoes committed:
chore: remove unneeded white space

petergoes committed:
refactor(compare-file-size): keep variable names consistant

petergoes committed:
style: add or remove extra tab

petergoes committed:
feat(logger): add updateLevel method to logger

The loglevel should be updateable during runtime

petergoes committed:
chore(package): add glob as a dependency

petergoes committed:
refactor(stats-reporter): generate table based on stats data

petergoes committed:
refactor(index): remove old stats module

petergoes committed:
refactor(reporter): readd line indicating where the files are stored

petergoes committed:
Merge branch 'master' into reporter-module

jbmoelker commented on bin/fastatic:
```diff
@@ -19,8 +20,8 @@ fastatic({
 		logLevel: program.logLevel
 	})
-	.then(config => logger.log('Done. Optimised static files in', config.dest))
+	.then(output => reporter(output))
```
Don't make it more generic then it needs to be:

``` javascript
.then(output => {
  const { filesizes, src, dest } = output;
  filesizeReporter({ filesizes, src, dest });
  // later: speedReporter({ ... })
})
```

jbmoelker commented on lib/reporter/index.js:
```diff
@@ -0,0 +1,14 @@
+const program = require('commander');
+const statsReporter = require('./stats-reporter');
+
+function reporter(output) {
```
this entire module can be dropped

jbmoelker commented on lib/reporter/stats-reporter.js:
```diff
@@ -0,0 +1,97 @@
+
+const logger = new Logger();
+
+function statsReporter(statsObj) {
```
rename to `filesizeReporter` as that more accurately describes what it does.

jbmoelker commented on lib/reporter/stats-reporter.js:
```diff
@@ -0,0 +1,97 @@
+const filesize = require('filesize');
+const Logger = require('../logger');
+const percent = require('calc-percent');
+const program = require('commander');
```
Don't use globals to pass values around. Rather pass `logLevel` or a `logger` instance as a parameter to this module.

If we have to make a change, consider turning `Logger` into a singleton, which can be required in different modules, while keeping it's configuration (like `logLevel`).

petergoes replied:
I created an issue for the logger

jbmoelker commented on lib/reporter/stats-reporter.js:
```diff
@@ -0,0 +1,97 @@
+	logger.log(table.toString());
+}
+
+function getSizesPerType(srcSizes, destSizes) {
```
rename to `calcSizesPerType`

jbmoelker commented on lib/reporter/stats-reporter.js:
```diff
@@ -0,0 +1,97 @@
+			type,
+			filesize(sizes.src),
+			filesize(sizes.dest),
+			filesize(sizes.saving) + ' (' + colors.green(percent(sizes.saving, [sizes.src], {suffix: '%'}))+ ')'
```
if you change

``` javascript
filesize(sizes.saving)
```

to

``` javascript
const sizes = tableData[type];
const saving = calcSaving(sizes.src, sizes.dest)
```

you don't need the more complicated `calculateSavingsPerType()` anymore.

petergoes committed:
chore: remove unused dependencies

jbmoelker commented on lib/reporter/stats-reporter.js:
```diff
@@ -0,0 +1,97 @@
+	}, {});
+}
+
+function generateTable(tableData) {
```
consistent naming: `createTable(data)`, `createTableRow(...)`

I also think there should be a function `createTableRow` (singular) instead of `getTableRows` (plural), so the `createTable` function would iterate through the data and call `createTableRow` for each.

petergoes committed:
refactor(stats-reporter): rename to filesizeReporter

jbmoelker commented on lib/reporter/stats-reporter.js:
```diff
@@ -0,0 +1,97 @@
+	const total = Object.keys(tableData).reduce((total, type) => {
+		total.src += tableData[type].src;
+		total.dest += tableData[type].dest;
+		total.saving += tableData[type].saving;
```
Like above `total.saving += calcSaving(total.src, total.dest);`.

petergoes committed:
chore(reporter): rename folder to reporters

petergoes committed:
refactor(fastatic): put filesizereporter module directly in fastatic

petergoes committed:
refactor(filesize-reporter): pass loglevel to reporter

petergoes committed:
refactor(filesize-reporter): calculate the savings in getSizesPerType

The dedicated calculateSavingsPerType function is a bit overkill

petergoes committed:
refactor(filesize-reporter): rename getSizesPerType function

petergoes committed:
refactor(filesize-reporter): use consistant naming

petergoes committed:
refactor(filesize-reporter): let createTable walk over data

Instead of a dedicated createTableRows method, combine the logic in the
createTable method

petergoes committed:
refactor(filesize-reporter): rename method for consistancy

petergoes committed:
refactor(filesize-reporter): make functions smaller

jbmoelker commented on bin/fastatic:
```diff
@@ -19,8 +20,12 @@ fastatic({
-	.then(config => logger.log('Done. Optimised static files in', config.dest))
+	.then(output => {
+		const { fileSize, dest } = output;
+		filesizeReporter(fileSize, program.logLevel);
```
Be consistent, with either `filesize` or `fileSize` (I think `filesize` is more common). You are mixing them in variable and function names.

The reporter is about filesizes, so it makes sense the first parameter is `filesize`. I can however imagine the reporter being extended at some point and then we may want to pass more options. So instead of passing `logLevel` directly as second argument I would change the signature to:
`filesizeReporter(filesizes, { logLevel: program.logLevel });`

petergoes replied:
I agree on the `filesize` / `fileSize` argument. However, the loglevel is temporarily. As soon as we make the logger a singleton, or the logger has the loglevel saved static, it can be removed here. The removal of the loglevel here will happen sooner than the addition of another argument. So lets leave it this way

jbmoelker commented on lib/reporters/filesize-reporter.js:
```diff
@@ -0,0 +1,56 @@
+
+const logger = new Logger();
+
+function filesizeReporter(statsObj, logLevel) {
```
I'd rename `statsObj` to `filesizes`

jbmoelker commented on lib/reporters/filesize-reporter.js:
```diff
@@ -0,0 +1,56 @@
+	return table;
+}
+
+function createHeaderData() {
```
rename to `getTableConfig()` or simply somewhere on top `const tableConfig = { ... }`

jbmoelker commented on lib/reporters/filesize-reporter.js:
```diff
@@ -0,0 +1,56 @@
+
+function createTable(srcObj, destObj) {
+	const table = new Table(createHeaderData());
+	const typeRows = Object.keys(srcObj).map(type => createTableRow(type, srcObj[type], destObj[type]));
```
I'd rename `typeRows` to `rows` and `totalRow` to `footer`.

jbmoelker commented on lib/reporters/filesize-reporter.js:
```diff
@@ -0,0 +1,56 @@
+	const typeRows = Object.keys(srcObj).map(type => createTableRow(type, srcObj[type], destObj[type]));
+	const totalRow = createTableRow('Total', calculateTotal(srcObj), calculateTotal(destObj));
+
+	[...typeRows, totalRow.map(item => colors.bold(item))]
```
I'd rewrite this to:

``` javascript
const rows = Object.keys(srcObj)
  .map(type => createTableRow(type, srcObj[type], destObj[type]))
  .filter(row => row.length);

const footer = createTableRow('Total', calculateTotal(srcObj), calculateTotal(destObj))
  .map(item => colors.bold(item)); // or maybe do this part in `createTotalRow()`

rows.map(row => table.push(row));
table.push(footer);
return table;
```

petergoes replied:
I left the pushing to the table the way it was. I think using the spread operator is making it more readable. I did however moved the `map` that makes the footer bold to the footer statement.

petergoes committed:
refactor(filesize-reporter): rename argument

petergoes committed:
refactor(filesize-reporter): rename variables, put tableConfig ontop

petergoes committed:
refactor(fastatic): keep var names consistant
